### PR TITLE
Napi rewrite

### PR DESCRIPTION
**Pull request by ralphtheninja**
### Why?

`N-API` is the latest interface for native addons in node.js. It's a set of c functions that lie between your native addon and v8. Main benefit is that your addon is independent of the v8 version, which means it will be compatible with _future_ versions of node.js.

So, as long as your native code doesn't change, it doesn't need to be recompiled as new versions of node.js are released. This helps greatly with both maintenance and also when delivering code to the users of your module.

The downside is a loss of performance since you're no longer interfacing directly with v8.

### Porting to N-API

There's been some work done already to port `leveldown` to `N-API` but I decided to give it a go myself. Mainly because accepting a huge PR as a maintainer means a lot of work, it's not much more work to actually do it yourself.

Also, personally, I've always felt that the v8 "lingo" and abstractions were difficult to understand. Basically, you need to understand how v8 works (to some extent) in order to write good native addons. Maybe it was just over my head.

### Design

However, I think "workers" as a concept is easy to understand, so I've tried to keep the code similar to the `nan` approach in this sense.

> But why all these C++ classes/structs? Why not use c structs? They are simpler!

`LevelDB` is a C++ library and the glue code was already written in C++. Also, I personally like virtual methods, which really come in handy when creating workers. You can abuse virtual methods as well, but if you keep it to a minimum (no multiple inheritance) it can be quite elegant.

Another reason for this was for simplicity during rewrite. The main focus when migrating from `nan` to `N-API` should be at the intersection of js and c++ land. The code in between should really not have to change that much.

> Why one binding.cc? The file has so many lines!

Correct. At first I wanted to keep a single file to make it easier during rewrite, but the more I thought of it, the more it made sense to keep _everything_ in one _amalgamated_ file. I'd prefer to keep it this way. It's so much easier to find things. Each async method has its worker right next to it.

### Feedback

> How can we help you review this?

Since this PR is quite big, I'd prefer to keep review at a bigger level. Please let me know if there's something in the code that's a big no no, e.g. if I'm using the wrong abstractions.

Input such as "Did you know you can do X instead of Y and then you don't have to Z?" is also very helpful.

Another important thing is that when you rewrite things and there are many details to keep in your head, you often forget things. So if you can find something that I have forgotten, that'd be great.

One thing that would be greatly needed is also help with finding bugs. It would be really helpful if you can make the code segfault. Also running benchmarks to compare with the previous implementation would be nice.

### Some stats

I tried running `test-for -n 100` for _pure_ tests (e.g. only the tape tests, without standard etc) and got the same average for both the `master` and the `napi` branch (~3.8s/test). Timing tests is most likely not a good measure, but it's a indication at least.

Total lines of native code in `src/`:

```
$ wc -l src/*
   32 src/async.h
   17 src/batch_async.cc
   26 src/batch_async.h
  148 src/batch.cc
   40 src/batch.h
   33 src/common.h
  290 src/database_async.cc
  168 src/database_async.h
  514 src/database.cc
  108 src/database.h
   86 src/iterator_async.cc
   43 src/iterator_async.h
  613 src/iterator.cc
   94 src/iterator.h
   38 src/leveldown_async.cc
   34 src/leveldown_async.h
   70 src/leveldown.cc
  117 src/leveldown.h
 2471 total
```

Compared to `binding.cc`:

```
$ wc -l binding.cc
1772
```

### CODE REVIEW

File: test/destroy-test.js [hunk 1]
```diff
@@ -37,8 +37,8 @@ test('test destroy non-existent directory', function (t) {
   rimraf(location, { glob: false }, function (err) {
     t.ifError(err, 'no error from rimraf()')
 
-    leveldown.destroy(location, function () {
-      t.is(arguments.length, 0, 'no arguments returned on callback')
+    leveldown.destroy(location, function (err) {
```
Comment (ralphtheninja) [inline test/destroy-test.js]: This behavior has changed. The callback now return `null`.

File: test/destroy-test.js [hunk 2]
```diff
@@ -55,8 +55,8 @@ test('test destroy non-existent parent directory', function (t) {
 
   t.notOk(fs.existsSync(parent), 'parent does not exist before')
 
-  leveldown.destroy(location, function () {
-    t.is(arguments.length, 0, 'no arguments returned on callback')
+  leveldown.destroy(location, function (err) {
```
Comment (ralphtheninja) [inline test/destroy-test.js]: Same here.

File: binding.gyp [hunk 3]
```diff
@@ -39,17 +39,10 @@
       "<(module_root_dir)/deps/leveldb/leveldb.gyp:leveldb"
     ],
     "include_dirs"  : [
-      "<!(node -e \"require('nan')\")"
+      "<!(node -e \"require('napi-macros')\")"
```
Comment (ralphtheninja) [inline binding.gyp]: :heart: @mafintosh

File: chained-batch.js [hunk 4]
```diff
@@ -1,25 +1,32 @@
 const util = require('util')
 const AbstractChainedBatch = require('abstract-leveldown').AbstractChainedBatch
+const binding = require('./binding')
 
 function ChainedBatch (db) {
   AbstractChainedBatch.call(this, db)
-  this.binding = db.binding.batch()
+  this.context = binding.batch_init(db.context)
```
Comment (ralphtheninja) [inline chained-batch.js]: Here's how we do it, regardless of where we are. We use an `_init` method to get a `context`.

File: chained-batch.js [hunk 5]
```diff
@@ -1,25 +1,32 @@
...
 
 ChainedBatch.prototype._write = function (options, callback) {
-  this.binding.write(callback)
+  // TODO (ensure docs covers the following)
+  // Note that we're passing in options here, which we didn't do before. We
+  // must do this so we can use the `sync` property, which we didn't handle before
+  // since we never passed in an object at time of creation (bug) (the previous c++
+  // used to assume we did this from the js side from the ChainedBatch()
+  // constructor).
+  binding.batch_write(this.context, options, callback)
```
Comment (ralphtheninja) [inline chained-batch.js]: This needs documentation. @vweevers Maybe write an issue for this before this is released?

File: package.json [hunk 6]
```diff
@@ -43,7 +44,7 @@
     "verify-travis-appveyor": "^3.0.0"
   },
   "scripts": {
-    "install": "prebuild-install || node-gyp rebuild",
```
Comment (ralphtheninja) [inline package.json]: After merge, I'll continue working on using `prebuildify` instead of `prebuild` and `prebuild-install`.

File: package.json [hunk 7]
```diff
@@ -43,7 +44,7 @@
...
+    "install": "node-gyp-build",
```
Comment (ralphtheninja) [inline package.json]: `node-gyp-build` checks if the module can be `require()`'d, if it fails it will recompile.

**Comment by ralphtheninja** [comment]
Picking some random people for review. It's fine if you don't have time or energy to review :smile:

File: chained-batch.js (hunk 5 above)
Comment (vweevers) [reply chained-batch.js]: #541

**Comment by vweevers** [comment]
> > Why one binding.cc? The file has so many lines!
>
> Correct. At first I wanted to keep a single file to make it easier during rewrite, but the more I thought of it, the more it made sense to keep everything in one amalgamated file. I'd prefer to keep it this way. It's so much easier to find things. Each async method has its worker right next to it.

I opened #542 for further discussion after this PR; personally I don't agree.

File: package.json (hunk 6 above)
Comment (vweevers) [reply package.json]: 👍 Folks, follow #482 for progress on prebuildify.

File: binding.cc [hunk 8]
```diff
@@ -542,63 +526,46 @@ struct Iterator {
...
-            } else if (lte_ != NULL) {
-              // TODO make a compount if statement
-              if (lte_->compare(keyStr) < 0) {
-                dbIterator_->Prev();
-              }
-            } else if (start_ != NULL) {
-              // TODO make a compount if statement
-              if (start_->compare(keyStr)) {
-                dbIterator_->Prev();
-              }
```
Comment (vweevers) [inline binding.cc]: The old code, though longer, made it more obvious that lt/lte/start are exclusive options.

**Comment by juliangruber** [comment]
Wow, what a herculean effort! ❤️

**Comment by vweevers** [comment]
There's a memory leak: running `node --expose-gc test/leak-tester.js` for a few minutes shows it growing to 1000% of initial memory usage.

```
getCount = 0 , putCount =  0 , rss = 101% 26M ["0","0","0","0","0","0","0"]
getCount = 1000 , putCount =  1000 , rss = 119% 31M ["0","0","0","0","0","0","0"]
..
getCount = 569000 , putCount =  553100 , rss = 1001% 262M ["4","13","126","423","0","0","0"]
```

File: binding.cc (hunk 8 above)
Comment (vweevers) [reply binding.cc]: Yes, let's.

**Comment by ralphtheninja** [comment]
@vweevers Which version of node are you running? Just making sure we use the same setup.

**Comment by vweevers** [comment]
It was 8.6.0 x64 on Windows. I also tried 8.14.0, doesn't exhibit the same growth rate. I'm running 10.14.1 now, currently at 3 million `gets` but holding steady at ~400% memory usage.

**Comment by ralphtheninja** [comment]
Hmm this is strange. I'm on v10.14.1 and I'm running up to 1000% even on the _master_ branch.

**Comment by vweevers** [comment]
> Hmm this is strange. I'm on v10.14.1 and I'm running up to 1000% even on the master branch.

Ai.

**Comment by ralphtheninja** [comment]
I think maybe we shouldn't look at percentage increase of rss memory. The baseline might be very different on windows compared to linux so 500% or 1500% is more based on how small/large the baseline is. What I _think_ we should focus on is to see if the total memory stabilizes around 2-3M gets.

**Comment by ralphtheninja** [comment]
node v11.4.0

master branch:

```
getCount = 3000000 , putCount =  2591736 , rss = 528% 179M ["2","29","165","1165","1367","0","0"]
getCount = 3001000 , putCount =  2592471 , rss = 541% 184M ["2","27","164","1165","1367","0","0"]
getCount = 3002000 , putCount =  2593203 , rss = 547% 186M ["3","25","163","1165","1367","0","0"]
getCount = 3003000 , putCount =  2593973 , rss = 546% 186M ["3","20","180","1165","1367","0","0"]
getCount = 3004000 , putCount =  2594699 , rss = 557% 189M ["4","18","185","1165","1367","0","0"]
getCount = 3005000 , putCount =  2595474 , rss = 562% 191M ["4","18","185","1165","1367","0","0"]
getCount = 3006000 , putCount =  2596220 , rss = 562% 191M ["4","17","187","1165","1367","0","0"]
getCount = 3007000 , putCount =  2596964 , rss = 555% 189M ["5","17","187","1165","1367","0","0"]
getCount = 3008000 , putCount =  2597725 , rss = 572% 195M ["5","17","187","1165","1367","0","0"]
getCount = 3009000 , putCount =  2598479 , rss = 547% 186M ["5","17","186","1165","1367","0","0"]
getCount = 3010000 , putCount =  2599220 , rss = 565% 192M ["6","17","185","1166","1367","0","0"]
getCount = 3011000 , putCount =  2599943 , rss = 575% 196M ["6","17","184","1167","1367","0","0"]
getCount = 3012000 , putCount =  2600674 , rss = 570% 194M ["6","17","183","1168","1367","0","0"]
getCount = 3013000 , putCount =  2601401 , rss = 572% 194M ["7","17","183","1168","1367","0","0"]
getCount = 3014000 , putCount =  2602139 , rss = 586% 199M ["7","17","181","1169","1367","0","0"]
getCount = 3015000 , putCount =  2602869 , rss = 592% 201M ["8","17","179","1171","1367","0","0"]
getCount = 3016000 , putCount =  2603613 , rss = 599% 204M ["8","16","174","1176","1367","0","0"]
getCount = 3017000 , putCount =  2604363 , rss = 579% 197M ["8","16","168","1169","1379","0","0"]
getCount = 3018000 , putCount =  2605098 , rss = 681% 232M ["9","16","167","1164","1385","0","0"]
getCount = 3019000 , putCount =  2605843 , rss = 519% 176M ["0","34","172","1164","1385","0","0"]
getCount = 3020000 , putCount =  2606578 , rss = 533% 181M ["0","34","172","1164","1385","0","0"]
getCount = 3021000 , putCount =  2607319 , rss = 529% 180M ["1","33","174","1164","1385","0","0"]
getCount = 3022000 , putCount =  2608061 , rss = 561% 191M ["1","31","176","1164","1385","0","0"]
getCount = 3023000 , putCount =  2608767 , rss = 560% 190M ["2","31","176","1164","1385","0","0"]
getCount = 3024000 , putCount =  2609500 , rss = 563% 192M ["2","29","177","1164","1385","0","0"]
getCount = 3025000 , putCount =  2610229 , rss = 557% 189M ["2","28","179","1164","1385","0","0"]
getCount = 3026000 , putCount =  2610980 , rss = 561% 191M ["2","28","179","1164","1385","0","0"]
getCount = 3027000 , putCount =  2611716 , rss = 565% 192M ["3","27","179","1164","1385","0","0"]
getCount = 3028000 , putCount =  2612430 , rss = 583% 198M ["3","26","180","1164","1385","0","0"]
getCount = 3029000 , putCount =  2613181 , rss = 571% 194M ["4","26","180","1164","1385","0","0"]
getCount = 3030000 , putCount =  2613919 , rss = 559% 190M ["4","24","181","1164","1385","0","0"]
getCount = 3031000 , putCount =  2614646 , rss = 581% 198M ["5","24","181","1164","1385","0","0"]
getCount = 3032000 , putCount =  2615399 , rss = 564% 192M ["5","23","183","1164","1385","0","0"]
getCount = 3033000 , putCount =  2616154 , rss = 579% 197M ["5","23","183","1164","1385","0","0"]
getCount = 3034000 , putCount =  2616884 , rss = 562% 191M ["6","21","183","1164","1385","0","0"]
getCount = 3035000 , putCount =  2617643 , rss = 552% 188M ["6","20","184","1164","1385","0","0"]
getCount = 3036000 , putCount =  2618391 , rss = 591% 201M ["6","20","184","1164","1385","0","0"]
getCount = 3037000 , putCount =  2619112 , rss = 584% 199M ["7","18","185","1164","1385","0","0"]
getCount = 3038000 , putCount =  2619833 , rss = 574% 195M ["7","17","186","1164","1385","0","0"]
getCount = 3039000 , putCount =  2620586 , rss = 580% 197M ["8","16","186","1165","1385","0","0"]
getCount = 3040000 , putCount =  2621318 , rss = 596% 203M ["8","15","181","1168","1385","0","0"]
```

napi branch:

```
getCount = 4149000 , putCount =  3396089 , rss = 560% 186M ["1","33","179","1228","2178","0","0"]
getCount = 4150000 , putCount =  3396742 , rss = 565% 188M ["1","32","179","1228","2178","0","0"]
getCount = 4151000 , putCount =  3397396 , rss = 575% 191M ["2","31","179","1228","2178","0","0"]
getCount = 4152000 , putCount =  3398060 , rss = 565% 188M ["2","30","179","1228","2178","0","0"]
getCount = 4153000 , putCount =  3398730 , rss = 562% 187M ["2","29","181","1228","2178","0","0"]
getCount = 4154000 , putCount =  3399370 , rss = 594% 197M ["3","29","181","1228","2178","0","0"]
getCount = 4155000 , putCount =  3400019 , rss = 588% 195M ["3","28","180","1228","2178","0","0"]
getCount = 4156000 , putCount =  3400653 , rss = 601% 200M ["3","27","180","1228","2178","0","0"]
getCount = 4157000 , putCount =  3401306 , rss = 590% 196M ["4","26","181","1228","2178","0","0"]
getCount = 4158000 , putCount =  3402010 , rss = 571% 190M ["4","25","181","1228","2178","0","0"]
getCount = 4159000 , putCount =  3402671 , rss = 602% 200M ["4","24","180","1228","2178","0","0"]
getCount = 4160000 , putCount =  3403305 , rss = 607% 202M ["5","24","180","1228","2178","0","0"]
getCount = 4161000 , putCount =  3403981 , rss = 591% 197M ["5","23","180","1228","2178","0","0"]
getCount = 4162000 , putCount =  3404638 , rss = 611% 203M ["5","23","180","1228","2178","0","0"]
getCount = 4163000 , putCount =  3405290 , rss = 624% 207M ["6","22","183","1228","2178","0","0"]
getCount = 4164000 , putCount =  3405967 , rss = 589% 196M ["6","21","183","1228","2178","0","0"]
getCount = 4165000 , putCount =  3406626 , rss = 627% 208M ["6","20","183","1228","2178","0","0"]
getCount = 4166000 , putCount =  3407298 , rss = 633% 210M ["7","19","183","1228","2178","0","0"]
getCount = 4167000 , putCount =  3407986 , rss = 619% 206M ["7","19","183","1228","2178","0","0"]
getCount = 4168000 , putCount =  3408652 , rss = 626% 208M ["7","18","185","1228","2178","0","0"]
getCount = 4169000 , putCount =  3409338 , rss = 642% 213M ["8","14","186","1229","2178","0","0"]
getCount = 4170000 , putCount =  3410003 , rss = 646% 215M ["8","14","183","1231","2178","0","0"]
getCount = 4171000 , putCount =  3410664 , rss = 641% 213M ["9","13","180","1234","2178","0","0"]
getCount = 4172000 , putCount =  3411315 , rss = 632% 210M ["9","13","175","1237","2178","0","0"]
getCount = 4173000 , putCount =  3411959 , rss = 621% 206M ["9","13","170","1242","2178","0","0"]
getCount = 4174000 , putCount =  3412619 , rss = 678% 225M ["10","13","167","1244","2178","0","0"]
getCount = 4175000 , putCount =  3413247 , rss = 719% 239M ["10","13","166","1245","2178","0","0"]
getCount = 4176000 , putCount =  3413922 , rss = 555% 184M ["0","37","166","1245","2178","0","0"]
getCount = 4177000 , putCount =  3414580 , rss = 557% 185M ["0","35","168","1245","2178","0","0"]
getCount = 4178000 , putCount =  3415252 , rss = 548% 182M ["1","34","168","1245","2178","0","0"]
getCount = 4179000 , putCount =  3415926 , rss = 557% 185M ["1","34","168","1245","2178","0","0"]
getCount = 4180000 , putCount =  3416599 , rss = 551% 183M ["1","31","172","1245","2178","0","0"]
getCount = 4181000 , putCount =  3417280 , rss = 548% 182M ["2","31","172","1245","2178","0","0"]
getCount = 4182000 , putCount =  3417956 , rss = 557% 185M ["2","29","173","1245","2178","0","0"]
getCount = 4183000 , putCount =  3418623 , rss = 546% 181M ["2","28","175","1245","2178","0","0"]
getCount = 4184000 , putCount =  3419279 , rss = 557% 185M ["3","27","176","1245","2178","0","0"]
```

Both seem to stabilize around 200Mb in total memory. I think we're good.

**Comment by vweevers** [comment]
> Both seem to stabilize around 200Mb in total memory. I think we're good.

👍 Agreed.

I'll open an issue for later, because I nevertheless want to try and stabilize the test itself (and write a bunch more tests while we're at it). In theory, any new allocated objects should be collected by `global.gc()` and that doesn't seem to be the case. There's various things we can do: add a warmup phase, disable certain LevelDB and V8 features, reuse the test's own callbacks and option objects, doing a put+get+del instead of creating a random key on every iteration, etc.

**Comment by ralphtheninja** [comment]
> I'll open an issue for later, because I nevertheless want to try and stabilize the test itself (and write a bunch more tests while we're at it). In theory, any new allocated objects should be collected by `global.gc()` and that doesn't seem to be the case. There's various things we can do: add a warmup phase, disable certain LevelDB and V8 features, reuse the test's own callbacks and option objects, doing a put+get+del instead of creating a random key on every iteration, etc.

Sounds great. I think we can do a lot here and hopefully also more generic work that can be used by other modules.

**Comment by ralphtheninja** [comment]
So we can get this merged?

**Comment by juliangruber** [comment]
If we merge this, shall we publish it as a new major, or at least a beta? This represents a significant enough change for me, that people might want to give it a try first.

**Comment by vweevers** [comment]
We already have a pending major. Making that a prerelease is a good idea.

**Comment by ralphtheninja** [comment]
Pre-major sounds good. Once that's ok we can start working on splitting things and making it purdier.

**Comment by vweevers** [comment]
Can we do some benchmarks?

**Comment by ralphtheninja** [comment]
@vweevers We have a `benchmarks` folder inside `levelup`. Gonna revive it a bit :) https://github.com/Level/levelup/tree/master/test/benchmarks

Maybe this should be a separate module?

**Comment by ralphtheninja** [comment]
Using `bench/db-bench.js` and plotting using `bench/db-bench-plot.sh` it seems that the napi version has _better_ throughput:

5mbench

**Comment by ralphtheninja** [comment]
Oh wait, it's the other way around. The left axis shows "Millseconds per write" and the right axis has "Throughput Mb/s". Then it makes more sense, since it's expected that napi has a trade off.

**Comment by mcollina** [comment]
cc @gabrielschulhof to check if there is anything to optimize here.

**Comment by ralphtheninja** [comment]
Second run. I _think_ I made a mistake the first time because I didn't remove the `build/` folder and since the napi branch uses `node-gyp-build` it didn't rebuild the binaries.

5mbench

I really need someone else to run this as well because this result is basically incredible if this is correct.

**Comment by vweevers** [comment]
An updated benchmark (#545) suggests napi is faster on Windows too! But I'll do some more runs after review of that PR.

**Comment by ralphtheninja** [comment]
There seem to be some variation but napi is _slightly_ faster. At least I think we can pretty confident that the performance haven't been degraded with respect to `db-bench.js`.

db-bench-20181215-00 43 39

**Comment by ralphtheninja** [comment]
Lets rebase again after https://github.com/Level/leveldown/pull/546 has been merged. If these benchmarks doesn't show anything unexpected, I think we should merge this PR.

**Comment by ralphtheninja** [comment]
And some more plots. Difficult to see because it's a lot of data (four files, each ~60Mb). The last one looks a bit funny in scale because of an extreme outlier.

write-random-20181215-12 59 03
write-sorted-20181215-12 54 09

**Comment by ralphtheninja** [comment]
@vweevers Question is now, do we squash this or make a merge commit? I'd personally like to merge and keep the rewrite history.

**Comment by vweevers** [comment]
> I'd personally like to merge and keep the rewrite history.

Sure 👍

**Comment by vweevers** [comment]
@ralphtheninja I think it's time for our perfect Neil Patrick Harris gif. I'll let you do the honors.

**Comment by ralphtheninja** [comment]
done

**Comment by ralphtheninja** [comment]
@vweevers I'm curious how the napi stuff works on node 6, e.g. the build https://travis-ci.org/Level/leveldown/jobs/468365583 

Has napi been backported to node 6?

**Pull request merged**